feat: add C implementation for `stats/base/dists/rayleigh/stdev`

DhruvArvindSingh opened the pull request:
Resolves #3810 .

## Description

> What is the purpose of this pull request?

This pull request:

-   This pull request adds C implementation for `@stdlib/stats/base/dists/rayleigh/stdev`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3810 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: add C implemenation for @stdlib/stats/base/dists/rayleigh/stdev

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DhruvArvindSingh committed:
fix: made changes in manifest.json

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/stats/base/dists/rayleigh/stdev.h"
+#include "stdlib/math/base/napi/unary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/README.md:
```diff
@@ -116,6 +116,88 @@ for ( i = 0; i < 10; i++ ) {
+    for ( i = 0; i < 25; i++ ) {
+        sigma = random_uniform( 0.0, 10.0 );
+        y = stdlib_base_dists_rayleigh_stdev( sigma );
+        printf( "σ: %lf, f(x;σ): %lf\n", sigma, y );
```
```suggestion
        printf( "σ: %lf, SD(X;σ): %lf\n", sigma, y );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/examples/c/example.c:
```diff
@@ -0,0 +1,38 @@
+	for ( i = 0; i < 25; i++ ) {
+		sigma = random_uniform( 0 , 10.0 );
+		y = stdlib_base_dists_rayleigh_stdev( sigma );
+		printf( "σ: %lf, stdev(X;σ): %lf\n", sigma, y );
```
```suggestion
		printf( "σ: %lf, SD(X;σ): %lf\n", sigma, y );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,139 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		sigma[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		sigma[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/src/main.c:
```diff
@@ -0,0 +1,57 @@
+/**
+* Returns the standard deviation of a Rayleigh distribution.
+*
+* @param {NonNegativeNumber} sigma - scale parameter
+* @returns {NonNegativeNumber} standard deviation
```
```suggestion
* @param sigma   scale parameter
* @return        standard deviation
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/src/main.c:
```diff
@@ -0,0 +1,57 @@
+* @param {NonNegativeNumber} sigma - scale parameter
+* @returns {NonNegativeNumber} standard deviation
+*
+* @example
+* double v = stdlib_base_dists_rayleigh_stdev( 9.0 );
+* // returns ~5.896
+*
+* @example
+* double v = stdlib_base_dists_rayleigh_stdev( 2.0 );
+* // returns ~1.31
+*
+* @example
+* double v = stdlib_base_dists_rayleigh_stdev( -0.2 );
+* // returns NaN
+*
+* @example
+* double v = stdlib_base_dists_rayleigh_stdev( NaN );
+* // returns NaN
```
```suggestion
* @example
* double v = stdlib_base_dists_rayleigh_stdev( 9.0 );
* // returns ~5.896
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/include/stdlib/stats/base/dists/rayleigh/stdev.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Returns the stdev for a Rayleigh distribution.
```
```suggestion
* Returns the standard deviation of a Rayleigh distribution.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/lib/native.js:
```diff
@@ -0,0 +1,57 @@
+
+/**
+* Returns the standard deviation of a Rayleigh distribution.
+*
+* @param {NonNegativeNumber} sigma - scale parameter
```
```suggestion
*
* @private
* @param {NonNegativeNumber} sigma - scale parameter
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/src/main.c:
```diff
@@ -0,0 +1,45 @@
+* // returns ~5.896
+*/
+double stdlib_base_dists_rayleigh_stdev( const double sigma ) {
+	const double SQRT4MPI = stdlib_base_sqrt( 4 - STDLIB_CONSTANT_FLOAT64_PI );
```
We can precompute this constant and define it in L25 as a `static const double`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/src/main.c:
```diff
@@ -0,0 +1,45 @@
+#include "stdlib/math/base/special/sqrt.h"
+#include "stdlib/constants/float64/pi.h"
+
+#define STDLIB_CONSTANT_FLOAT64_SQRT2 1.4142135623730950488
```
Should be defined as a `static const double`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/manifest.json:
```diff
@@ -0,0 +1,84 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/constants/float64/sqrt-two",
```
This dependency is not used currently. Once we precompute the constant as suggested, `sqrt` and `pi` will become unnecessary here and below, too.

Planeshifter committed:
chore: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/README.md:
```diff
@@ -116,6 +116,88 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_rayleigh_stdev( sigma )
+
+Evaluates the stdev for a Rayleigh distribution.
```
```suggestion
Returns the [standard deviation][standard-deviation] of a [Rayleigh][rayleigh-distribution] distribution with scale `sigma`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/stdev/README.md:
````diff
@@ -116,6 +116,88 @@ for ( i = 0; i < 10; i++ ) {
+        printf( "σ: %lf, SD(X;σ): %lf\n", sigma, y );
+    }
+}
+```
````
````suggestion
</section>

<!-- /.usage -->

<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

</section>

<!-- /.notes -->

<!-- C API usage examples. -->

<section class="examples">

### Examples

```c
#include "stdlib/stats/base/dists/rayleigh/stdev.h"
#include <stdlib.h>
#include <stdio.h>

static double random_uniform( const double min, const double max ) {
    double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
    return min + ( v*(max-min) );
}

int main( void ) {
    double sigma;
    double y;
    int i;

    for ( i = 0; i < 25; i++ ) {
        sigma = random_uniform( 0.0, 10.0 );
        y = stdlib_base_dists_rayleigh_stdev( sigma );
        printf( "σ: %lf, SD(X;σ): %lf\n", sigma, y );
    }
}
```
````

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
bench: use uniform to draw random numbers

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
Should be ready now.
